transformer_engine/pytorch/float8_tensor.py

```diff
@@ -540,14 +526,6 @@ def to_dtype(self, dtype: torch.dtype) -> Float8Tensor:
             dtype=dtype,
         )
 
-    def _reset_caches(self) -> None:
-        """Reset cached values
-
-        Should be called after any in-place operation.
-
-        """
-        self._transpose = None
```

Comment:
Removing the automatic cache clearing makes using the transpose cache a much more manual and dangerous process. Consider something like:
```python
matmul(x, w.transpose(0, 1))
w -= learning_rate * w.grad
matmul(x, w.transpose(0, 1))
```
Previously we could just set `update_cache="lazy"`. Now there needs to be manual logic to figure out the microbatch step, or else it would provide the stale values.

Reply:
In this example, caching is not used, so a fresh transpose will be computed each time.

Reply:
If caching is used, it is reasonable to expect the user to know when to reuse a cached value and when to force recompute. This is consistent with our design of `is_first_microbatch` argument to the forward for module APIs.

Reply:
**Note:** we use 2 args `cache` and `update_cache` to support this logic.

Reply:
I think we're overfitting to our particular use-case. For example, in https://github.com/NVIDIA/TransformerEngine/pull/707 I want to pass `Float8Tensor`s between ops as inputs or dgrads:
```
class DbiasCastTranspose:
    def backward(self, dy):
        db = dy.sum(dim=0)
        dx = cast_transpose(dy)  # Creates Float8Tensor with transpose cache

class FP8Linear:
    def backward(self, dy):
        if not isinstance(dy, Float8Tensor):
           dy = Float8Tensor.to_float8(dy)
        dx = Float8Tensor(...)  # No transpose cache
        fp8_gemm(w.transpose()._data, dy.transpose()._data, out=dx._data)
        return dx
```
`Linear` has no idea where its input came from. Maybe it's from `DbiasCastTranspose` (`Float8Tensor` with cached transpose), `PureFP8Linear` (`Float8Tensor` without cached transpose), or a non-FP8 op. Note that FP8 attention is like `FP8Linear`. Our current approach gives us a lot of flexibility and I think we should abandon it only when really necessary.